### DISCUSSION TITLE
Websocket example: 400 error and ENOENT in send_file

### DISCUSSION THREAD

**Discussion by Jarshvor** [miguelgrinberg/microdot#191] [Q&A] [answered]
I am trying to get the websocket example working on my ESP32.

But I get the following error when navigating to the default 'root' route:
```
>>> import websocket_echo
Traceback (most recent call last):
  File "/lib/microdot/microdot.py", line 1326, in dispatch_request
  File "/lib/microdot/microdot.py", line 40, in invoke_handler
  File "websocket_echo.py", line 9, in index
  File "/lib/microdot/microdot.py", line 793, in send_file
OSError: [Errno 2] ENOENT
```
When navigating to '/echo' I get nothing in the repl but 400 error in the browser.

There seems to be some problem with the send_file method?
But I cannot figure out what file is apparently missing that it throws that ENOENT error.

My file tree looks something like this:
```
├── lib
│   ├── microdot
│   │   ├── __init__.py
│   │   ├── microdot.py
│   │   ├── utemplate.py
│   │   └── websocket.py
│   └── utemplate
│       ├── compiled.py
│       ├── recompile.py
│       └── source.py
├── templates
│   ├── index.html
│   └── index_html.py
└── websocket_echo.py
```

And the files are copies of:
[microdot.py](https://github.com/miguelgrinberg/microdot/blob/main/src/microdot/microdot.py)
[webocket.py](https://github.com/miguelgrinberg/microdot/blob/main/src/microdot/websocket.py)
[index.html](https://github.com/miguelgrinberg/microdot/blob/main/examples/websocket/index.html)
[websocket_echo.py](https://github.com/miguelgrinberg/microdot/blob/main/examples/websocket/echo.py)

**Comment by miguelgrinberg** [comment, marked as answer]
The `ENOENT` error is short for "no such file or directory". It means that the path that you are passing to `send_file()` does not represent a file that can be opened. To me it sounds like your application is confusing templates with static files. The `send_file()` function is used to send static files, but you have your index.html file in the templates directory.

The quick and dirty solution would be to change the `send_file()` call to `send_file('templates/index.html')`. This function does not have any notion of where files are stored, you have to give it an absolute or relative path to the file. As a general recommendation I would suggest that if your html file is not going to be a template that you put it in a different directory to avoid confusion with templates, which are a special kind of file. Often used directory names for static files are `static` or `public`.

**Reply by Jarshvor** [reply]
Thank you Miguel.
I managed to run the provided `websocket_echo.py `example successfully once I changed the location of the `index.html` file.

I was, in fact, confusing the static files with template files. Thank you for calling my attention on that.
I've now taken note that I have to provide a relative or absolute path to the send_file function.

I'm still pretty new to all this so thanks for bearing with me.
